pkg/ring/lifecycler.go

```diff
@@ -497,6 +498,12 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 			return ringDesc, true, nil
 		}
 
+		// If the ingester failed to clean it's ring entry up in can leave it's state in LEAVING.
+		// Move it into ACTIVE to ensure the ingester joins the ring.
+		if ingesterDesc.State == LEAVING && len(ingesterDesc.Tokens) != 0 {
+			ingesterDesc.State = ACTIVE
```

Comment:
We should not set it to `ACTIVE` till we have confirmed that we have enough tokens as set in the config.

Reply:
(I think so looking at the existing code, not sure if it is the requirement)

Reply:
I am not sure if PENDING would be correct, as pending ingesters can be contacted to take over tokens from another ingester. We could use JOINING state, but that requires extra code to call autoJoin from this state (or using observe period flag).

I think ACTIVE is fine, if we don't care too much about number of tokens (I don't think it's a big deal) or it hasn't changed between restarts.

Reply:
I've updated it to check for NumTokens instead of just non-zero amount of tokens. Though I'm not sure it would make much of a difference.

Reply:
Given that we don't change number of tokens often, this should not make a difference. If you increase the tokens between rollout, this might need a different handling (but not sure if we need to handle that in this PR).